src/store/index-level.ts

```diff
@@ -37,75 +33,124 @@ export class IndexLevel {
+      // 1. creates a unique entry of the property-value pair per data/object
+      // 2. when we need to delete all indexes of a given data ID (`delete()`), we can reconstruct the index keys and remove the indexes efficiently
+      //
+      // example keys (\u0000 is just shown for illustration purpose because it is the delimiter used to join the string segments below):
+      // 'interface\u0000"Records"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
+      // 'method\u0000"Write"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
+      // 'schema\u0000"http://ud4kyzon6ugxn64boz7v"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
+      // 'dataCid\u0000"bafkreic3ie3cxsblp46vn3ofumdnwiqqk4d5ah7uqgpcn6xps4skfvagze"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
+      // 'dateCreated\u0000"2023-05-25T18:23:29.425008Z"\u0000bafyreigs3em7lrclhntzhgvkrf75j2muk6e7ypq3lrw3ffgcpyazyw6pry'
+      const key = this.join(propertyName, this.encodeValue(value), id);
+      operations.push({ type: 'put', key, value: id });
     }
-    ops.push({ type: 'put', key: `__${id}__prefixes`, value: JSON.stringify(prefixes) });
 
-    return this.db.batch(ops, options);
+    // create a reverse lookup entry for data ID -> its indexes
+    // this is for indexes deletion (`delete()`): so that given the data ID, we are able to delete all its indexes
+    // we can consider putting this info in a different data partition if this ever becomes more complex/confusing
+    operations.push({ type: 'put', key: `__${id}__indexes`, value: JSON.stringify(indexes) });
+
+    return this.db.batch(operations, options);
   }
 
   async query(filter: Filter, options?: IndexLevelOptions): Promise<Array<string>> {
-    const requiredProperties = new Set<string>();
-    const missingPropertiesForID: { [id: string]: Set<string> } = { };
-    const promises: Promise<Matches>[] = [ ];
-    const matchedIDs: string[] = [ ];
-
-    async function checkMatches(property: string, promise: Promise<Matches>): Promise<void> {
-      promises.push(promise);
-
-      for (const [ _, id ] of await promise) {
-        missingPropertiesForID[id] ??= new Set<string>([ ...requiredProperties ]);
-        missingPropertiesForID[id].delete(property);
-        if (missingPropertiesForID[id].size === 0) {
-          matchedIDs.push(id);
-        }
-      }
-    }
+    // Note: We have an array of Promises in order to support OR (anyOf) matches when given a list of accepted values for a property
+    const propertyNameToPromises: { [key: string]: Promise<string[]>[] } = {};
 
+    // Do a separate DB query for each property in `filter`
+    // We will find the union of these many individual queries later.
     for (const propertyName in filter) {
       const propertyFilter = filter[propertyName];
 
-      if (typeof propertyFilter === 'object' && propertyFilter !== null) {
+      if (propertyFilter === null) {
```

Comment:
tbh I don't think we even need this check. Typescript doesn't complain if I delete it.

Reply:
Removing.